aten/src/ATen/native/Gesv.h

```diff
@@ -1,30 +1,92 @@
+#include <utility>
 #include "ATen/ATen.h"
 
 namespace at { namespace native {
 
-static inline void checkInputs(const Tensor& self, const Tensor& A) {
-  if (A.size(-1) != A.size(-2)) {
-    AT_ERROR("A must be batches of square matrices, "
-        "but they are %lld by %lld matrices",
-        (long long)A.size(-1), (long long)A.size(-2));
+static inline bool isTransposeContiguous(Tensor& self) {
+  return self.dim() == 2 &&
+          self.stride(0) == 1 &&
+          self.stride(1) == self.size(0);
+}
+
+/* gesv takes (self, A) and returns (sol, lu).
+ * (i)  output tensors (sol, lu) may be same as input tensors (self, A)
+ * (ii) for 2D matrices, .t_() represents their column-major format
+ *
+ * Cases:
+ * 1) `out` has correct shape but elements do not form a contiguous
+ * chunk of memory. Since shape is correct, we don't resize_ it. Instead, we
+ * clone the input tensor into a buffer, use the buffer for Lapack and finally
+ * copy the buffer to the output tensor.
+ *
+ * 2) out.t() is contiguous:
+ *    (i)  &in == &out: use out.data() as is. Do nothing
+ *    (ii) &in != &out: copy in.t() to out.t()
+ * 3) out.t() is not contiguous:
+ *    - resize_ should fix contiguity/size issues
+ *    (i)  &in == &out: copy in.t().clone() to out (same tensor)
+ *    (ii) &in != &out: copy in.t() to out
+ */
+static inline void prepareIOTensors(
+    const Tensor& in, Tensor& out, Tensor& temp,
+    int64_t& x, int64_t& y) {
```

Comment:
So the function takes in `in` and returns `out`, `temp`, `x`, `y`. To my understanding, the output tensor always (sometimes?) has size (y, x): is it necessary to return both `x` and `y`?

Reply:
It's just easier to pass them by reference here so that I don't have to get (y, x) from the output tensor (which may not have .size(1), so I might have to do `dim == 1 ? 1 : .size(1)` again. I don't really use the values written for `ax, ay` (`A`'s dimensions), but pass `bx`, `by` to Lapack.

Reply:
Actually, I agree -- it'll probably be better to just get `bx`, `by` in the `AT_DISPATCH` block and use it instead of returning from the helper. Will change this in the next commit

Reply:
Does the output tensor always have size (y, x)? Sorry, I'm a little confused by the logic

Reply:
Yeah, the output tensor will be 2D and have size (y, x) after the prepare call. If `temp_sol` is being used, `sol` was already the correct shape, and in the case where it is 1D it will be resized to be 2D, so we can always do:
```
AT_DISPATCH_FLOATING_TYPES(...
    int64_t bx = sol.size(1);
    int64_t by = sol.size(0);
    ...
})